### ISSUE TITLE
Providing patroni with a custom CA bundle

### ISSUE THREAD

**Issue by melnik13** [zalando/postgres-operator#1877]
Hello,

Is that possible to provide Patroni with a custom CA bundle for interacting with Kubernetes API Server?

By default it uses `/var/run/secrets/kubernetes.io/serviceaccount/ca.crt` which is set by Kuberneres (it consists the data from the `Secret` referred by the main `ServiceAccount`). My problem is that my Kuberneres' certificate is signed by an intermediate certificate which is signed by a self-signed root certificate. Alas, Patroni (or, to be exact, the SSL library that is being used by Patroni) couldn't establish the TLS-session with the Kubernetes API Server, as `/var/run/secrets/kubernetes.io/serviceaccount/ca.crt` consists only a "shortened" version of my CA chain (only the intermediate certificate is included).

Is there a way to ask Patroni to use the full version of my CA chain when it's interacting with the Kubernetes API Server? Of course, I know that that's possible to set `kubernetes.cacert` in the Patroni's configuration file to make it using any file instead of the `ca.crt` provided by Kuberneres, but I can't find the way to accomplish that by Postgres Operator.

Would anyone be so kind as to help me to find a way to do that?

Thanks in advance!

**Comment by melnik13** [comment]
Thank you for the hint!

It might be that I do something wrong, but I have to admit that the environment variables don't change anything.

Here's a snippet from the file containing the configuration values for Helm:
```
spec:                                                                                                                                                                                         env:                                                                                                                                                                                          - name: PATRONI_KUBERNETES_CACERT
      value: /tls/ca.crt
    - name: PATRONI_LOG_LEVEL
      value: DEBUG
    - name: PATRONI_LOG_TRACEBACK_LEVEL
      value: DEBUG
```

As I can see, they are applied, here's what I see in the description of the 1st pod:
```
Containers:
  postgres:
    Environment:
      PATRONI_KUBERNETES_CACERT:    /tls/ca.crt
      PATRONI_LOG_LEVEL:            DEBUG
      PATRONI_LOG_TRACEBACK_LEVEL:  DEBUG
```

But at the same time, I still get the same error messages in the output of Patroni:
```
2022-05-12 00:48:19,237 WARNING: Retrying (Retry(total=0, connect=None, read=None, redirect=0, status=None)) after connection broken by 'SSLError(SSLError(1, '[SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed (_ssl.c:852)'),)': /api/v1/namespaces/default/endpoints/kubernetes
2022-05-12 00:48:19,248 ERROR: Failed to get "kubernetes" endpoint from https://10.233.0.1:443: MaxRetryError("HTTPSConnectionPool(host='10.233.0.1', port=443): Max retries exceeded with url: /api/v1/namespaces/default/endpoints/kubernetes (Caused by SSLError(SSLError(1, '[SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed (_ssl.c:852)'),))",)
```

At the same time, `/tls/ca.crt` contains the full CA certification chain. I'm totally sure in that, because everything works fine when I edit `/home/postgres/postgres.yml` (adding the `kubernetes.cacert` parameter contatning `/tls/ca.crt`) and restart Patroni.

By the way `PATRONI_LOG_LEVEL` seems to be ignored too, because I don't see any messages lower than the `INFO` level in the Patroni's output.

That's also pretty odd that the other environment variables of this pod doesn't contain the `PATRONI_` prefix in their names. Here's the full list of the environment variables that I see in the pod's description:
```
Containers:
  postgres:
    Environment:
      SCOPE:                        acid-kloudster-backend
      PGROOT:                       /home/postgres/pgdata/pgroot
      POD_IP:                        (v1:status.podIP)
      POD_NAMESPACE:                postgresql (v1:metadata.namespace)
      PGUSER_SUPERUSER:             postgres
      KUBERNETES_SCOPE_LABEL:       cluster-name
      KUBERNETES_ROLE_LABEL:        spilo-role
      PGPASSWORD_SUPERUSER:         <set to the key 'password' in secret 'postgres.acid-kloudster-backend.credentials.postgresql.acid.zalan.do'>  Optional: false
      PGUSER_STANDBY:               standby
      PGPASSWORD_STANDBY:           <set to the key 'password' in secret 'standby.acid-kloudster-backend.credentials.postgresql.acid.zalan.do'>  Optional: false
      PAM_OAUTH2:                   https://info.example.com/oauth2/tokeninfo?access_token= uid realm=/employees
      HUMAN_ROLE:                   zalandos
      PGVERSION:                    14
      KUBERNETES_LABELS:            {"application":"spilo"}
      SPILO_CONFIGURATION:          {"postgresql":{},"bootstrap":{"initdb":[{"auth-host":"md5"},{"auth-local":"trust"}],"users":{"zalandos":{"password":"","options":["CREATEDB","NOLOGIN"]}},"dcs":{}}}
      DCS_ENABLE_KUBERNETES_API:    true
      PATRONI_KUBERNETES_CACERT:    /tls/ca.crt
      PATRONI_LOG_LEVEL:            DEBUG
      PATRONI_LOG_TRACEBACK_LEVEL:  DEBUG
      SSL_CERTIFICATE_FILE:         /tls/tls.crt
      SSL_PRIVATE_KEY_FILE:         /tls/tls.key
      SSL_CA_FILE:                  /tls/ca.crt
```
Why does not these variables' names contain the `PATRONI_` prefix?

Thanks.

**Comment by dmvolod** [comment]
Sorry, my bad, Zalando operator don't run Partoni directly and utilizes Spilo image for it
Please have a look at the following documentation to choose correct env variable for the certificate
https://github.com/zalando/spilo/blob/master/ENVIRONMENT.rst

**Comment by dmvolod** [comment]
Oh, yes, true. For this case, some fixes in the Spilo should be implemented to propagate these options to the Patroni

**Comment by howels** [comment]
Is this resolved via https://github.com/zalando/spilo/releases/tag/2.1-p6 or other recent changes?

**Comment by melnik13** [comment]
> Is this resolved via https://github.com/zalando/spilo/releases/tag/2.1-p6 or other recent changes?

Haven't tried that yet, because I decided that using a "long" CA bundle for a Kubernetes wasn't a good idea at all. There are way too many issues with other components.

**Comment by D1StrX** [comment]
@khmarochos Just trying to understand the scope of "many issues with other components", do you mean also other stuff than just Zalando Postgresql? Having the exact same issue here, regarding a custom k8s CA.

**Comment by khmarochos** [comment]
@D1StrX, right, a chained CA causes lots of pain. I had similar problems with the Elastic stack. I also have a feeling that there were something else, though I can't remember what exactly. So, I decided to amandon the idea to use chained CAs for Kubernetes clusters.
I've found a couple of links to the questions I raised before:
 * https://serverfault.com/questions/1100038/kubernetes-using-an-intermediate-ca-which-certificate-is-signed-by-a-self-signe,
 * https://github.com/kubernetes-sigs/kubespray/issues/8242.

There are not too much of feedback, so it might be that there are not too many perverts to do that. :-)

**Comment by lenglet-k** [comment]
hello, it's possible to have a status of this issue ? It's possible to custom ca bundle for kubernetes in patroni configuration ?

**Comment by hydramyst** [comment]
hello this my config to fix issue
```
  env:
    - name: KUBERNETES_CACERT
      value: /etc/conf/ca-bundle.crt
  additionalVolumes:
    - mountPath: /etc/conf/ca-bundle.crt
      name: conf
      subPath: ca-certificates.pem
      targetContainers:
        - all
      volumeSource:
        configMap:
          name: trusted-ca-bundle
```
